common/reflect_resource.go

```diff
@@ -419,6 +428,11 @@ func typeToSchema(v reflect.Value, aliases map[string]string) map[string]*schema
 	return scm
 }
 
+func maxRecursionDepthExceeded(typeField reflect.StructField, timesVisited map[string]int) bool {
```

Comment:
implementation is spread over multiple places, making it more fragile. add a small new type to contain all the functionality:

```go
type recursionTracking struct { 
  visited map[string]int
  maxDepth int 
}
func (rt recursionTracking) DepthExceeded(typeField reflect.StructField) bool
func (rt recursionTracking) Copy() recursionTracking
func (rt recursionTracking) name(typeField reflect.StructField) string
```